### PR TITLE
support query edges of vertex in page

### PR DESCRIPTION
**Pull request by zhoney**
implemented: #658

Change-Id: I26e8efb1e8e3547f403dd97e2ddc367df69eae51

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/QueryList.java [hunk 1]
```diff
@@ -271,10 +272,21 @@ public PageIterator(Iterator<BackendEntry> iterator, String page) {
             this.page = page;
         }
 
+        public PageIterator(Iterator<BackendEntry> iterator,
+                            PageState pageState) {
+            this.iterator = iterator;
+            this.page = pageState.toString();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/QueryList.java]: just update page type to PageState

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/QueryList.java [hunk 2]
```diff
@@ -271,10 +272,21 @@ public PageIterator(Iterator<BackendEntry> iterator, String page) {
...
+        }
+
         public Iterator<BackendEntry> iterator() {
             return this.iterator;
         }
 
+        public boolean hasNullPage() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/QueryList.java]: update to hasNextPage

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageIds.java [hunk 3]
```diff
@@ -36,6 +37,11 @@ public PageIds(Set<Id> ids, String page) {
         this.page = page;
     }
 
+    public PageIds(Set<Id> ids, PageState pageState) {
+        this.ids = ids;
+        this.page = pageState.toString();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageIds.java]: update page type to PageState

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java [hunk 4]
```diff
@@ -124,39 +124,35 @@ public HugeVertexStep(final VertexStep<E> originVertexStep) {
...
-        for (ConditionQuery q : queries) {
-            // Query by sort-keys
-            if (withEdgeCond && edgeLabels.length > 0) {
-                TraversalUtil.fillConditionQuery(conditions, q, graph);
-                if (!GraphTransaction.matchEdgeSortKeys(q, graph)) {
-                    // Can't query by sysprop and by index (HugeGraph-749)
-                    q.resetUserpropConditions();
-                }
+        // Query by sort-keys
+        ConditionQuery q = query;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java]: delete var q

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java [hunk 5]
```diff
@@ -124,39 +124,35 @@ public HugeVertexStep(final VertexStep<E> originVertexStep) {
...
-            if (withEdgeCond) {
-                results.extend(TraversalUtil.filterResult(conditions, edges));
-            } else {
-                results.extend(edges);
-            }
+        // Do filter by edge conditions
+        if (withEdgeCond) {
+            return TraversalUtil.filterResult(conditions, edges);
+        } else {
+            return edges;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java]: remove else-statement and return edges if not withEdgeCond

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/TraversalUtil.java [hunk 6]
```diff
@@ -587,10 +588,21 @@ public static void retriveSysprop(List<HasContainer> hasContainers,
...
         E.checkState(holder != null,
                      "Invalid paging traversal: %s", traversal.getClass());
-        return (String) holder.metadata(PageInfo.PAGE);
+        Object page = holder.metadata(PageInfo.PAGE);
+        if (page instanceof PageInfo) {
+            // Traversal with condition like: g.V().has("x", 1).has("~page", "")
+            return page.toString();
+        } else if (page instanceof PageState) {
+            // Traversal without condition like: g.V().has("~page", "")
+            return ((PageState) page).page();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/TraversalUtil.java]: also call toString()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageIds.java [hunk 7]
```diff
@@ -27,27 +27,26 @@
 public final class PageIds {
 
     public static final PageIds EMPTY = new PageIds(ImmutableSet.of(),
-                                                    (String) null);
+                                                    (PageState) null);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageIds.java]: PageState. EMPTY

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageInfo.java [hunk 8]
```diff
@@ -101,10 +101,26 @@ public static PageInfo fromBytes(byte[] bytes) {
...
+    /**
+     * Only used for vertex/edge scan
+     * @param iterator iterator of scan query
+     * @return PageInfo string
+     */
+    public static String pageInfo(Iterator<?> iterator) {
+        E.checkState(iterator instanceof Metadatable,
+                     "Invalid paging iterator: %s", iterator.getClass());
+        Object page = ((Metadatable) iterator).metadata(PAGE);
+        E.checkState(page == null || page instanceof PageInfo,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageInfo.java]: PageInfo won't be null

Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/PageInfo.java]: null is returned if offset > total for PageInfo

**Pull request merged**